File: typescript/src/extras/encoders.ts
import { ByteMode, FuqrError, type Encoder, type Mode } from "../fuqr.ts";

export const NumericMode: Mode = {
  accepts: (byte) => 0x30 <= byte && byte <= 0x39,
  cci: (version) => 10 + (version > 9 ? 2 : 0) + (version > 26 ? 2 : 0),
  // 10 bits per 3 digits, 4 or 7 bits for 1 or 2 leftover digits
  bitLen: (len, version) => 4 + NumericMode.cci(version) + Math.ceil((len * 10) / 3),
  encode(bytes, start, end, version, push) {
    push(0b0001, 4);
    push(end - start, NumericMode.cci(version));
    let i = start;
    for (; i + 3 <= end; i += 3) {
      push((bytes[i] - 0x30) * 100 + (bytes[i + 1] - 0x30) * 10 + (bytes[i + 2] - 0x30), 10);
    }
    switch (end - i) {
      case 1:
        push(bytes[i] - 0x30, 4);
        break;
      case 2:
        push((bytes[i] - 0x30) * 10 + (bytes[i + 1] - 0x30), 7);
        break;
    }
  },
};

// Alphanumeric value of each byte, or 255 if not alphanumeric
const B45 = new Uint8Array(256).fill(255);
for (let i = 0; i < 45; i++) {
  B45["0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:".charCodeAt(i)] = i;
}

export const AlphanumericMode: Mode = {
  accepts: (byte) => byte < 256 && B45[byte] !== 255,
  cci: (version) => 9 + (version > 9 ? 2 : 0) + (version > 26 ? 2 : 0),
  // 11 bits per 2 chars, 6 bits for 1 leftover char
  bitLen: (len, version) => 4 + AlphanumericMode.cci(version) + Math.ceil((len * 11) / 2),
  encode(bytes, start, end, version, push) {
    push(0b0010, 4);
    push(end - start, AlphanumericMode.cci(version));
    let i = start;
    for (; i + 2 <= end; i += 2) {
      push(B45[bytes[i]] * 45 + B45[bytes[i + 1]], 11);
    }
    if (i < end) push(B45[bytes[i]], 6);
  },
};

// Content is stored as char codes, which match UTF-8 bytes for valid content
function charCodes(content: string, mode: Mode, name: string) {
  const bytes = new Uint8Array(content.length);
  for (let i = 0; i < content.length; i++) {
    const byte = content.charCodeAt(i);
    if (!mode.accepts(byte)) {
      throw new FuqrError("INVALID_ENCODING", `Content is not ${name}`);
    }
    bytes[i] = byte;
  }
  return bytes;
}

export class NumericEncoder implements Encoder {
  public bytes: Uint8Array;
  constructor(content: string) {
    this.bytes = charCodes(content, NumericMode, "numeric");
  }

  bitLen(version: number) {
    return NumericMode.bitLen(this.bytes.length, version);
  }

  encode(version: number, push: (bits: number, len: number) => void) {
    NumericMode.encode(this.bytes, 0, this.bytes.length, version, push);
  }
}

export class AlphanumericEncoder implements Encoder {
  static byteToB45(c: number): number {
    return c < 256 ? B45[c] : 255;
  }

  public bytes: Uint8Array;
  constructor(content: string) {
    this.bytes = charCodes(content, AlphanumericMode, "alphanumeric");
  }

  bitLen(version: number) {
    return AlphanumericMode.bitLen(this.bytes.length, version);
  }

  encode(version: number, push: (bits: number, len: number) => void) {
    AlphanumericMode.encode(this.bytes, 0, this.bytes.length, version, push);
  }
}

// Encodes bytes as the given segments, which should cover every byte in order
export class SegmentEncoder implements Encoder {
  public bytes: Uint8Array;
  public segments: { mode: Mode; start: number; end: number }[];
  constructor(bytes: Uint8Array, segments: { mode: Mode; start: number; end: number }[]) {
    for (const { mode, start, end } of segments) {
      if (start < 0 || end < start || bytes.length < end) {
        throw new FuqrError("INVALID_ENCODING", `Segment ${start}..${end} is out of bounds`);
      }
      for (let i = start; i < end; i++) {
        if (!mode.accepts(bytes[i])) {
          throw new FuqrError("INVALID_ENCODING", `Byte ${i} is not valid in its segment`);
        }
      }
    }
    this.bytes = bytes;
    this.segments = segments;
  }

  bitLen(version: number) {
    let bits = 0;
    for (const { mode, start, end } of this.segments) bits += mode.bitLen(end - start, version);
    return bits;
  }

  encode(version: number, push: (bits: number, len: number) => void) {
    for (const { mode, start, end } of this.segments) {
      mode.encode(this.bytes, start, end, version, push);
    }
  }
}

type Segment = { mode: number; start: number; end: number };

// Indexed by MixedEncoder segment mode
const MODES = [NumericMode, AlphanumericMode, ByteMode];

// Cheapest mode each byte fits in: 0 numeric, 1 alphanumeric, 2 byte
// All multibyte UTF-8 bytes look like 1xxx_xxxx, so they are always 2
const MODE = new Uint8Array(256);
for (let i = 0; i < 256; i++) {
  MODE[i] = NumericMode.accepts(i) ? 0 : AlphanumericMode.accepts(i) ? 1 : 2;
}

export class MixedEncoder implements Encoder {
  public bytes: Uint8Array;
  public modes: Uint8Array;
  public segments: Segment[];
  public version: number;
  // Segmentation only depends on char count indicator lengths, so there are
  // only 3 distinct results: versions 1-9, 10-26, 27-40
  private cached: { bits: number; segments: Segment[] }[];

  constructor(content: string) {
    this.segments = [];
    this.version = 0;
    this.cached = [];

    const bytes = new TextEncoder().encode(content);
    const modes = new Uint8Array(bytes.length);
    for (let i = 0; i < bytes.length; i++) modes[i] = MODE[bytes[i]];
    this.bytes = bytes;
    this.modes = modes;
  }

  bitLen(version: number): number {
    const group = version < 10 ? 0 : version < 27 ? 1 : 2;
    const result = (this.cached[group] ??= this.segment(version));
    this.segments = result.segments;
    this.version = version;
    return result.bits;
  }

  // Costs are in sixths of a bit so each char has a fixed cost:
  // numeric 20, alphanumeric 33, byte 48. Segments round up when closed.
  private segment(version: number) {
    const modes = this.modes;
    const n = modes.length;

    if (n === 0) {
      return {
        bits: ByteMode.bitLen(0, version),
        segments: [{ mode: 2, start: 0, end: 0 }],
      };
    }

    // One mode throughout is optimal as a single segment
    let mode = modes[0];
    let i = 1;
    while (i < n && modes[i] === mode) i++;
    if (i === n) {
      return {
        bits: MODES[mode].bitLen(n, version),
        segments: [{ mode, start: 0, end: n }],
      };
    }

    // header + first char
    const start0 = (4 + NumericMode.cci(version)) * 6 + 20;
    const start1 = (4 + AlphanumericMode.cci(version)) * 6 + 33;
    const start2 = (4 + ByteMode.cci(version)) * 6 + 48;

    // prev[i] packs the mode at i - 1 for each mode at i, 2 bits per mode
    const prev = new Uint8Array(n);

    let c0 = mode === 0 ? start0 : Infinity;
    let c1 = mode <= 1 ? start1 : Infinity;
    let c2 = start2;

    for (i = 1; i < n; i++) {
      // Cost of closing a segment in each mode
      const r0 = Math.ceil(c0 / 6) * 6;
      const r1 = Math.ceil(c1 / 6) * 6;
      const r2 = Math.ceil(c2 / 6) * 6;
      mode = modes[i];
      let p = 0;

      // Staying wins ties to avoid pointless segments
      // Switching to the same mode is never better than staying
      if (mode === 0) {
        const from = r1 <= r2 ? 1 : 2;
        const swap = (from === 1 ? r1 : r2) + start0;
        const stay = c0 + 20;
        if (stay <= swap) {
          c0 = stay;
        } else {
          c0 = swap;
          p = from;
        }
      } else {
        c0 = Infinity;
      }

      if (mode <= 1) {
        const from = r0 <= r2 ? 0 : 2;
        const swap = (from === 0 ? r0 : r2) + start1;
        const stay = c1 + 33;
        if (stay <= swap) {
          c1 = stay;
          p |= 1 << 2;
        } else {
          c1 = swap;
          p |= from << 2;
        }
      } else {
        c1 = Infinity;
      }

      const from = r0 <= r1 ? 0 : 1;
      const swap = (from === 0 ? r0 : r1) + start2;
      const stay = c2 + 48;
      if (stay <= swap) {
        c2 = stay;
        p |= 2 << 4;
      } else {
        c2 = swap;
        p |= from << 4;
      }

      prev[i] = p;
    }

    let m = c0 <= c1 ? (c0 <= c2 ? 0 : 2) : c1 <= c2 ? 1 : 2;
    const bits = Math.ceil((m === 0 ? c0 : m === 1 ? c1 : c2) / 6);

    const segments: Segment[] = [];
    let end = n;
    for (let i = n - 1; i >= 1; i--) {
      const p = (prev[i] >> (m * 2)) & 0b11;
      if (p !== m) {
        segments.push({ mode: m, start: i, end });
        end = i;
        m = p;
      }
    }
    segments.push({ mode: m, start: 0, end });
    segments.reverse();

    return { bits, segments };
  }

  encode(version: number, push: (bits: number, len: number) => void) {
    this.bitLen(version);
    for (const { mode, start, end } of this.segments) {
      MODES[mode].encode(this.bytes, start, end, version, push);
    }
  }
}

File: rust/src/encoders.rs
use crate::{ByteMode, Encoder, FuqrError, Mode, Version};

const fn cci_diff(version: Version) -> u8 {
    (version > 9) as u8 * 2 + (version > 26) as u8 * 2
}

#[derive(Debug, Clone, Copy)]
pub struct NumericMode;
impl Mode for NumericMode {
    fn indicator(&self) -> u16 {
        0b0001
    }
    fn cci_len(&self, version: Version) -> u8 {
        10 + cci_diff(version)
    }
    // 10 bits per 3 digits, 4 or 7 bits for 1 or 2 leftover digits
    fn seg_len(&self, len: usize, version: Version) -> usize {
        4 + self.cci_len(version) as usize + (len * 10).div_ceil(3)
    }
}

#[derive(Debug, Clone, Copy)]
pub struct AlphanumericMode;
impl Mode for AlphanumericMode {
    fn indicator(&self) -> u16 {
        0b0010
    }
    fn cci_len(&self, version: Version) -> u8 {
        9 + cci_diff(version)
    }
    // 11 bits per 2 chars, 6 bits for 1 leftover char
    fn seg_len(&self, len: usize, version: Version) -> usize {
        4 + self.cci_len(version) as usize + (len * 11).div_ceil(2)
    }
}

#[derive(Debug)]
pub struct NumericEncoder<'a> {
    pub bytes: &'a [u8],
}
impl<'a> NumericEncoder<'a> {
    pub fn new(content: &'a str) -> Result<Self, FuqrError> {
        let bytes = content.as_bytes();
        if !bytes.iter().all(u8::is_ascii_digit) {
            return Err(FuqrError::InvalidEncoding {
                message: "Content is not numeric",
            });
        }
        Ok(Self { bytes })
    }
}
impl<'a> Encoder for NumericEncoder<'a> {
    fn bit_len(&mut self, version: Version) -> usize {
        NumericMode.seg_len(self.bytes.len(), version)
    }
    fn encode(&mut self, version: Version, push: &mut dyn FnMut(u16, u8)) {
        let bytes = self.bytes;

        push(NumericMode.indicator(), 4);
        push(bytes.len() as u16, NumericMode.cci_len(version));
        let mut groups = bytes.chunks_exact(3);
        for g in &mut groups {
            push(
                (g[0] - b'0') as u16 * 100 + (g[1] - b'0') as u16 * 10 + (g[2] - b'0') as u16,
                10,
            );
        }
        match *groups.remainder() {
            [a] => push((a - b'0') as u16, 4),
            [a, b] => push((a - b'0') as u16 * 10 + (b - b'0') as u16, 7),
            _ => {}
        }
    }
}

// Alphanumeric value of each byte, or 255 if not alphanumeric
const B45: [u8; 256] = {
    let chars = b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:";
    let mut table = [255; 256];
    let mut i = 0;
    while i < chars.len() {
        table[chars[i] as usize] = i as u8;
        i += 1;
    }
    table
};

#[derive(Debug)]
pub struct AlphanumericEncoder<'a> {
    pub bytes: &'a [u8],
}
impl<'a> AlphanumericEncoder<'a> {
    pub fn new(content: &'a str) -> Result<Self, FuqrError> {
        let bytes = content.as_bytes();
        if bytes.iter().any(|&b| B45[b as usize] == 255) {
            return Err(FuqrError::InvalidEncoding {
                message: "Content is not alphanumeric",
            });
        }
        Ok(Self { bytes })
    }
    pub fn byte_to_b45(c: u8) -> u8 {
        B45[c as usize]
    }
}
impl<'a> Encoder for AlphanumericEncoder<'a> {
    fn bit_len(&mut self, version: Version) -> usize {
        AlphanumericMode.seg_len(self.bytes.len(), version)
    }
    fn encode(&mut self, version: Version, push: &mut dyn FnMut(u16, u8)) {
        let bytes = self.bytes;

        push(AlphanumericMode.indicator(), 4);
        push(bytes.len() as u16, AlphanumericMode.cci_len(version));
        let mut pairs = bytes.chunks_exact(2);
        for p in &mut pairs {
            push(
                B45[p[0] as usize] as u16 * 45 + B45[p[1] as usize] as u16,
                11,
            );
        }
        if let [a] = *pairs.remainder() {
            push(B45[a as usize] as u16, 6);
        }
    }
}

// Cheapest mode each byte fits in: 0 numeric, 1 alphanumeric, 2 byte
// All multibyte UTF-8 bytes look like 1xxx_xxxx, so they are always 2
const MODE: [u8; 256] = {
    let mut table = [2; 256];
    let mut i = 0;
    while i < 256 {
        if B45[i] < 10 {
            table[i] = 0;
        } else if B45[i] != 255 {
            table[i] = 1;
        }
        i += 1;
    }
    table
};

fn seg_len(mode: u8, len: usize, version: Version) -> usize {
    match mode {
        0 => NumericMode.seg_len(len, version),
        1 => AlphanumericMode.seg_len(len, version),
        _ => ByteMode.seg_len(len, version),
    }
}

// Versions 1-9, 10-26, 27-40 have different char count indicator lengths
fn cci_group(version: Version) -> usize {
    (version > 9) as usize + (version > 26) as usize
}

const UNKNOWN: usize = usize::MAX;
const INF: usize = usize::MAX / 4;

/// Encodes content in the shortest mix of numeric, alphanumeric and byte segments.
///
/// `scratch` must be at least as long as the content in UTF-8 bytes. It holds
/// the chosen mode of each byte.
#[derive(Debug)]
pub struct MixedEncoder<'a> {
    pub bytes: &'a [u8],
    modes: &'a mut [u8],
    // Segmentation only depends on char count indicator lengths, so there are
    // only 3 distinct results, one per cci_group
    bits: [usize; 3],
    // Which group's segmentation `modes` currently holds
    filled: usize,
}
impl<'a> MixedEncoder<'a> {
    pub fn new(content: &'a str, scratch: &'a mut [u8]) -> Self {
        let bytes = content.as_bytes();
        assert!(
            scratch.len() >= bytes.len(),
            "scratch buffer shorter than content"
        );
        Self {
            bytes,
            modes: &mut scratch[..bytes.len()],
            bits: [UNKNOWN; 3],
            filled: UNKNOWN,
        }
    }

    fn fill(&mut self, version: Version) {
        let group = cci_group(version);
        if self.filled != group {
            self.bits[group] = self.segment(version);
            self.filled = group;
        }
    }

    // Costs are in sixths of a bit so each char has a fixed cost:
    // numeric 20, alphanumeric 33, byte 48. Segments round up when closed.
    fn segment(&mut self, version: Version) -> usize {
        let bytes = self.bytes;
        let modes = &mut *self.modes;
        let n = bytes.len();

        if n == 0 {
            return ByteMode.seg_len(0, version);
        }

        // One mode throughout is optimal as a single segment
        let first = MODE[bytes[0] as usize];
        if bytes.iter().all(|&b| MODE[b as usize] == first) {
            modes.fill(first);
            return seg_len(first, n, version);
        }

        // header + first char
        let start0 = (4 + NumericMode.cci_len(version) as usize) * 6 + 20;
        let start1 = (4 + AlphanumericMode.cci_len(version) as usize) * 6 + 33;
        let start2 = (4 + ByteMode.cci_len(version) as usize) * 6 + 48;

        let mut c0 = if first == 0 { start0 } else { INF };
        let mut c1 = if first <= 1 { start1 } else { INF };
        let mut c2 = start2;

        // modes[i] packs the mode at i - 1 for each mode at i, 2 bits per mode
        for i in 1..n {
            let mode = MODE[bytes[i] as usize];

            // Only byte mode is possible, so nothing to compare
            if mode == 2 && c0 == INF && c1 == INF {
                c2 += 48;
                modes[i] = 2 << 4;
                continue;
            }

            // Cost of closing a segment in each mode
            let r0 = c0.div_ceil(6) * 6;
            let r1 = c1.div_ceil(6) * 6;
            let r2 = c2.div_ceil(6) * 6;
            let mut p = 0;

            // Staying wins ties to avoid pointless segments
            if mode == 0 {
                let (from, r) = if r1 <= r2 { (1, r1) } else { (2, r2) };
                let stay = c0 + 20;
                if stay <= r + start0 {
                    c0 = stay;
                } else {
                    c0 = r + start0;
                    p = from;
                }
            } else {
                c0 = INF;
            }

            if mode <= 1 {
                let (from, r) = if r0 <= r2 { (0, r0) } else { (2, r2) };
                let stay = c1 + 33;
                if stay <= r + start1 {
                    c1 = stay;
                    p |= 1 << 2;
                } else {
                    c1 = r + start1;
                    p |= from << 2;
                }
            } else {
                c1 = INF;
            }

            let (from, r) = if r0 <= r1 { (0, r0) } else { (1, r1) };
            let stay = c2 + 48;
            if stay <= r + start2 {
                c2 = stay;
                p |= 2 << 4;
            } else {
                c2 = r + start2;
                p |= from << 4;
            }

            modes[i] = p;
        }

        let (mut m, cost) = if c0 <= c1 {
            if c0 <= c2 {
                (0, c0)
            } else {
                (2, c2)
            }
        } else if c1 <= c2 {
            (1, c1)
        } else {
            (2, c2)
        };

        // Replace packed modes with the chosen mode of each byte
        for i in (1..n).rev() {
            let p = (modes[i] >> (m * 2)) & 0b11;
            modes[i] = m;
            m = p;
        }
        modes[0] = m;

        cost.div_ceil(6)
    }
}
impl<'a> Encoder for MixedEncoder<'a> {
    fn bit_len(&mut self, version: Version) -> usize {
        let group = cci_group(version);
        if self.bits[group] == UNKNOWN {
            self.fill(version);
        }
        self.bits[group]
    }

    fn encode(&mut self, version: Version, push: &mut dyn FnMut(u16, u8)) {
        self.fill(version);

        let bytes = self.bytes;
        let modes = &*self.modes;
        let n = bytes.len();

        if n == 0 {
            push(ByteMode.indicator(), 4);
            push(0, ByteMode.cci_len(version));
            return;
        }

        let mut start = 0;
        while start < n {
            let mode = modes[start];
            let mut end = start + 1;
            while end < n && modes[end] == mode {
                end += 1;
            }
            let segment = &bytes[start..end];

            match mode {
                0 => {
                    push(NumericMode.indicator(), 4);
                    push(segment.len() as u16, NumericMode.cci_len(version));
                    let mut groups = segment.chunks_exact(3);
                    for g in &mut groups {
                        push(
                            (g[0] - b'0') as u16 * 100
                                + (g[1] - b'0') as u16 * 10
                                + (g[2] - b'0') as u16,
                            10,
                        );
                    }
                    match *groups.remainder() {
                        [a] => push((a - b'0') as u16, 4),
                        [a, b] => push((a - b'0') as u16 * 10 + (b - b'0') as u16, 7),
                        _ => {}
                    }
                }
                1 => {
                    push(AlphanumericMode.indicator(), 4);
                    push(segment.len() as u16, AlphanumericMode.cci_len(version));
                    let mut pairs = segment.chunks_exact(2);
                    for p in &mut pairs {
                        push(
                            B45[p[0] as usize] as u16 * 45 + B45[p[1] as usize] as u16,
                            11,
                        );
                    }
                    if let [a] = *pairs.remainder() {
                        push(B45[a as usize] as u16, 6);
                    }
                }
                _ => {
                    push(ByteMode.indicator(), 4);
                    push(segment.len() as u16, ByteMode.cci_len(version));
                    for &b in segment {
                        push(b as u16, 8);
                    }
                }
            }

            start = end;
        }
    }
}
